shared-bindings/displayio/Display.c

```diff
@@ -356,17 +373,140 @@ const mp_obj_property_t displayio_display_bus_obj = {
+    static const mp_arg_t allowed_args[] = {
+      { MP_QSTR_x, MP_ARG_INT | MP_ARG_KW_ONLY, {.u_int = -1} },
+        { MP_QSTR_y, MP_ARG_INT | MP_ARG_KW_ONLY, {.u_int = -1} },
+        { MP_QSTR_width, MP_ARG_INT | MP_ARG_KW_ONLY, {.u_int = -1} },
+        { MP_QSTR_height, MP_ARG_INT | MP_ARG_KW_ONLY, {.u_int = -1} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    displayio_display_obj_t *self = native_display(pos_args[0]);
+    mp_int_t x = args[ARG_x].u_int;
+    mp_int_t y = args[ARG_y].u_int;
+    mp_int_t w = args[ARG_width].u_int;
+    mp_int_t h = args[ARG_height].u_int;
+
+    uint16_t buffer_size = 128; // In uint32_ts
+    displayio_area_t area = {
+      .x1 = x,
+      .y1 = y,
+      .x2 = x + w,
+      .y2 = y + h
+    };
+    displayio_area_t clipped;
+    // Clip the area to the display by overlapping the areas. If there is no overlap then we're done.
+    if (!displayio_display_clip_area(self, &area, &clipped)) {
+        return  mp_const_none;
+    }
+    uint16_t subrectangles = 1;
+    uint16_t rows_per_buffer = displayio_area_height(&clipped);
+    uint8_t pixels_per_word = (sizeof(uint32_t) * 8) / self->colorspace.depth;
+    uint16_t pixels_per_buffer = displayio_area_size(&clipped);
+    if (displayio_area_size(&clipped) > buffer_size * pixels_per_word) {
+        rows_per_buffer = buffer_size * pixels_per_word / displayio_area_width(&clipped);
+        if (rows_per_buffer == 0) {
+            rows_per_buffer = 1;
+        }
+        // If pixels are packed by column then ensure rows_per_buffer is on a byte boundary.
+        if (self->colorspace.depth < 8 && !self->colorspace.pixels_in_byte_share_row) {
+            uint8_t pixels_per_byte = 8 / self->colorspace.depth;
+            if (rows_per_buffer % pixels_per_byte != 0) {
+                rows_per_buffer -= rows_per_buffer % pixels_per_byte;
+            }
+        }
+        subrectangles = displayio_area_height(&clipped) / rows_per_buffer;
+        if (displayio_area_height(&clipped) % rows_per_buffer != 0) {
+            subrectangles++;
+        }
+        pixels_per_buffer = rows_per_buffer * displayio_area_width(&clipped);
+        buffer_size = pixels_per_buffer / pixels_per_word;
+        if (pixels_per_buffer % pixels_per_word) {
+            buffer_size += 1;
+        }
+    }
+
+    // Allocated and shared as a uint32_t array so the compiler knows the
+    // alignment everywhere.
+    uint32_t buffer[buffer_size];
+    volatile uint32_t mask_length = (pixels_per_buffer / 32) + 1;
+    uint32_t mask[mask_length];
+
+    for (uint16_t k = 0; k < mask_length; k++) {
+      mask[k] = 0x00000000;
+    }
+    for (uint16_t k = 0; k < buffer_size; k++) {
+      buffer[k] = 0x00000000;
+    }
+
+    displayio_display_fill_area(self, &area, mask, buffer);
+
+    mp_obj_array_t *result = array_new(BYTEARRAY_TYPECODE, buffer_size * 4);
```

Comment:
Instead of allocating an array, please accept a bytearray or array in as an argument so that the buffer can be reused. Reusing the buffer will make it easier to allocate and reduce fragmentation.